client/lib/media/actions.js

```diff
@@ -99,91 +100,139 @@ MediaActions.add = function( siteId, files ) {
+						assign( transientMedia, {
+							URL: blobUrl,
+							guid: blobUrl
+						} );
+					} );
+				}
+
+				// Generate from window.File object
+				const fileUrl = window.URL.createObjectURL( file );
+				assign( transientMedia, {
+					URL: fileUrl,
+					guid: fileUrl
+				} );
+				return Promise.resolve();
 			};
-		}
-
-		debug( 'Uploading media to %d from %o', siteId, file );
-		return lastUpload.then( () => {
-			// Achieve series upload by waiting for the previous promise to
-			// resolve before starting this item's upload
-			const action = { type: 'RECEIVE_MEDIA_ITEM', id, siteId };
-			return wpcom.site( siteId )[ addHandler ]( {}, file ).then( ( data ) => {
-				Dispatcher.handleServerAction( Object.assign( action, {
-					data: data.media[ 0 ]
-				} ) );
-				// also refetch media limits
-				Dispatcher.handleServerAction( {
-					type: 'FETCH_MEDIA_LIMITS',
-					siteId: siteId
+
+			return createTransientMedia().then( () => {
+				Dispatcher.handleViewAction( {
+					type: 'CREATE_MEDIA_ITEM',
+					siteId: siteId,
+					data: transientMedia
+				} );
+
+				media.push( {
+					file: file,
+					transientMedia: transientMedia
 				} );
-			} ).catch( ( error ) => {
-				Dispatcher.handleServerAction( Object.assign( action, { error } ) );
 			} );
 		} );
-	}, Promise.resolve() );
+
+		return Promise.all( mediaPromises ).then( () => media );
+	};
+
+	const uploadMedia = ( media ) => {
+		return media.reduce( ( lastUpload, item ) => {
+			// Abort upload if file fails to pass validation.
+			if ( MediaValidationStore.getErrors( siteId, item.transientMedia.ID ).length ) {
+				return Promise.resolve();
+			}
+
+			let file = item.file;
+
+			// Determine upload mechanism by object type
+			const isUrl = 'string' === typeof file;
+			const addHandler = isUrl ? 'addMediaUrls' : 'addMediaFiles';
+
+			// Assign parent ID if currently editing post
+			const post = PostEditStore.get();
+
+			if ( post && post.ID && ! isPlainObject( file ) ) {
+				file = {
+					parent_id: post.ID,
+					[ isUrl ? 'url' : 'file' ]: file
+				};
+			}
+
+			debug( 'Uploading media to %d from %o', siteId, file );
+			return lastUpload.then( () => {
+				// Achieve series upload by waiting for the previous promise to
+				// resolve before starting this item's upload
+				const action = { type: 'RECEIVE_MEDIA_ITEM', id: item.transientMedia.ID, siteId };
+
+				return wpcom.site( siteId )[ addHandler ]( {}, file )
+					.then( data => {
+						Dispatcher.handleServerAction( Object.assign( action, {
+							data: data.media[ 0 ]
+						} ) );
+						// also refetch media limits
+						Dispatcher.handleServerAction( {
+							type: 'FETCH_MEDIA_LIMITS',
+							siteId: siteId
+						} );
+					} )
+					.catch( error => {
+						Dispatcher.handleServerAction( Object.assign( action, { error } ) );
+					} );
+			} );
+		}, Promise.resolve() );
+	};
+
+	const create = createMedia();
+
+	if ( options.waitForUpload ) {
```

Comment:
I wanted to be consistent with what existed before. Previously, the final promise resolved when the uploading was complete. The tests use `waitForUpload`.